python/lib/core/dmod/core/data_domain_detectors.py

```diff
@@ -0,0 +1,468 @@
+            subclass's registration ``name``.
+
+        Returns
+        -------
+        DataDomain
+            The detected domain.
+
+        Raises
+        ------
+        ValueError
+            Raised if a :class:`DataFormat` appears multiple times across both ``excluded_formats`` and
+            ``suggested_formats``; i.e., if any data format value is duplicated in the hypothetical list produced by
+            ``list(kwargs.get('excluded_formats', [])) + list(kwargs.get('suggested_formats', []))``.
+        DmodRuntimeError
+            If it was not possible to properly detect the domain.
+        """
+        def try_detection(d_format: DataFormat) -> Optional[DataDomain]:
+            subclasses = ItemDataDomainDetectorRegistry.get_instance().get_for_format(d_format)
+            if 'sort_key' in kwargs:
+                subclasses = sorted(subclasses, key=kwargs['sort_key'])
+            for subclass_type in subclasses:
+                try:
+                    return subclass_type(item=self._item, item_name=self._item_name).detect()
+                except:
+                    pass
+            return None
+
+        excluded = kwargs.get('excluded_formats', set())
+        if isinstance(excluded, DataFormat):
+            excluded = {excluded}
+        suggested = kwargs.get('suggested_formats', list())
+        if isinstance(suggested, DataFormat):
+            suggested = [suggested]
+        if not excluded.isdisjoint(suggested):
+            raise ValueError(f"Can't include data format in both exclusions and suggestions for domain detection.")
+        if len(suggested) != len(set(suggested)):
+            raise ValueError(f"Can't include data format multiple times in ordered suggestions for domain detection.")
+
+        remaining_formats = {df for df in DataFormat if df not in excluded}
+
+        # Try suggestions first, returning immediately if any are successful
+        for data_format in suggested:
+            remaining_formats.remove(data_format)
+            result = try_detection(d_format=data_format)
+            if result is not None:
+                return result
+
+        # Now we get to others
+        main_trials = (try_detection(d_format=df) for df in remaining_formats)
+        main_results = [t for t in main_trials if t is not None]
+        if len(main_results) == 0:
+            raise DmodRuntimeError("No domain could be detected for item.")
+        elif len(main_results) == 1:
+            return main_results[0]
+        # Multiple results mean there's a problem (also, they can't be equal because they will have different formats)
+        else:
+            raise DmodRuntimeError(f"Multiple conflicting domain detected for item in the following formats: "
+                                   f"{','.join([d.data_format.name for d in main_results])}")
+
+
+# Register the universal tracker type
+ItemDataDomainDetectorRegistry.get_instance().register(UniversalItemDomainDetector)
```

Comment:
I think you are probably right, if we are not doing automatic registration (i.e., what I was trying to do originally).  But this actually has be rethinking moving away from that, as it feels counterintuitive.

Instead of that, I am probably going to move away from what I have currently to implementing the universal detector closer to the types it depends on in _modeldata_.  It can then be clear from the implementation what is important and what other detector subclasses it has access to.  Things can always build on that at the package level if needed, but otherwise they don't have to do anything other than the typical import-and-use.

I will have to figure out how to deal with _modeldata_ not being a strict dependency of _client_ though.

Reply:
As mentioned above, I'm just going to cave for now and make it a strict dependency.  If/when we discover that causing a real problem, we can come back to doing something differently.